// @phero/core/dist/tsx/Literal.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Literal = void 0;
const typescript_1 = __importDefault(require("typescript"));
class Literal {
    static get null() {
        return typescript_1.default.factory.createNull();
    }
    static get undefined() {
        return typescript_1.default.factory.createIdentifier("undefined");
    }
    static number(value) {
        return typescript_1.default.factory.createNumericLiteral(value);
    }
    static string(value) {
        return typescript_1.default.factory.createStringLiteral(value);
    }
    static get true() {
        return typescript_1.default.factory.createTrue();
    }
    static get false() {
        return typescript_1.default.factory.createFalse();
    }
    static boolean(value) {
        return value ? typescript_1.default.factory.createTrue() : typescript_1.default.factory.createFalse();
    }
    static object(...props) {
        return typescript_1.default.factory.createObjectLiteralExpression(props);
    }
    static array(...elements) {
        return typescript_1.default.factory.createArrayLiteralExpression(elements);
    }
    static type(...members) {
        return typescript_1.default.factory.createTypeLiteralNode(members);
    }
    static function(props) {
        return typescript_1.default.factory.createFunctionTypeNode(props.typeParams, props.params, props.type);
    }
    static regularExpression(text) {
        return typescript_1.default.factory.createRegularExpressionLiteral(text);
    }
}
exports.Literal = Literal;
