// @phero/server/dist/commands/export/compileExportToJS.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@phero/core");
const typescript_1 = __importDefault(require("typescript"));
function compileExportToJS(files) {
    const printer = typescript_1.default.createPrinter({
        newLine: typescript_1.default.NewLineKind.LineFeed,
        removeComments: true,
    });
    const vHost = new core_1.VirtualCompilerHost({
        declaration: false,
    });
    for (const file of files) {
        const tsSourceFile = typescript_1.default.createSourceFile(file.name, "", typescript_1.default.ScriptTarget.ES5, false, typescript_1.default.ScriptKind.TS);
        const tsContent = printer.printList(typescript_1.default.ListFormat.SourceFileStatements, typescript_1.default.factory.createNodeArray(file.nodes), tsSourceFile);
        vHost.addFile(file.name, tsContent);
    }
    const prog = vHost.createProgram(files.filter((f) => f.isRoot).map((f) => f.name));
    if (prog.emit().diagnostics.length) {
        throw new Error("Export build doesn't compile");
    }
    return files.map((file) => {
        const jsFileName = file.name.replace(".ts", ".js");
        return {
            name: jsFileName,
            content: vHost.getFile(jsFileName),
        };
    });
}
exports.default = compileExportToJS;
