// @phero/server/dist/commands/export/vercel/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const compileExportToJS_1 = __importDefault(require("../compileExportToJS"));
const generateLibFile_1 = __importDefault(require("./generateLibFile"));
const generateServiceHandlerFile_1 = require("./generateServiceHandlerFile");
const generateServiceIndexFile_1 = __importDefault(require("./generateServiceIndexFile"));
function generateVercelExport(app, metaExportFiles) {
    const bundles = app.services.flatMap((service) => ({
        name: `.vercel/output/functions/${service.name}/[func].func`,
        files: [
            ...(0, compileExportToJS_1.default)([
                {
                    name: "lib.ts",
                    nodes: (0, generateLibFile_1.default)(),
                    isRoot: true,
                },
                {
                    name: "handler.ts",
                    nodes: (0, generateServiceHandlerFile_1.generateServiceHandlerFile)(service),
                    isRoot: true,
                },
                {
                    name: "index.ts",
                    nodes: (0, generateServiceIndexFile_1.default)(),
                    isRoot: true,
                },
            ]),
            {
                name: ".vc-config.json",
                content: `{
    "runtime": "nodejs16.x",
    "handler": "index.js",
    "maxDuration": 3,
    "launcherType": "Nodejs",
    "shouldAddHelpers": true,
    "shouldAddSourcemapSupport": true
}`,
            },
            ...Object.entries(metaExportFiles).map(([name, content]) => ({ name, content })),
        ],
    }));
    const otherFiles = [
        {
            name: ".vercel/output/config.json",
            content: JSON.stringify({
                version: 3,
                routes: app.services.map((service) => ({
                    src: `^/${service.name}/(?<func>[^/]+?)(?:/)?$`,
                    dest: `/${service.name}/[func]?func=$func`,
                })),
            }),
        },
    ];
    return { bundles, otherFiles };
}
exports.default = generateVercelExport;
