// @phero/core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualCompilerHost = exports.tsx = exports.parsePheroApp = exports.generateNonModelParser = exports.generateModelParser = exports.generateRPCProxy = exports.generateAppDeclarationFile = exports.ReferenceMaker = exports.generateNamespace = exports.generateModel = exports.generateClientFunction = exports.getDeclarationForVersion = exports.isModel = exports.parseAppDeclarationFileContent = void 0;
var parseAppDeclaration_1 = require("./parseAppDeclaration");
Object.defineProperty(exports, "parseAppDeclarationFileContent", { enumerable: true, get: function () { return parseAppDeclaration_1.parseAppDeclarationFileContent; } });
Object.defineProperty(exports, "isModel", { enumerable: true, get: function () { return parseAppDeclaration_1.isModel; } });
var ParsedAppDeclarationVersion_1 = require("./ParsedAppDeclarationVersion");
Object.defineProperty(exports, "getDeclarationForVersion", { enumerable: true, get: function () { return ParsedAppDeclarationVersion_1.getDeclarationForVersion; } });
var code_gen_lib_1 = require("./code-gen-lib");
Object.defineProperty(exports, "generateClientFunction", { enumerable: true, get: function () { return code_gen_lib_1.generateClientFunction; } });
Object.defineProperty(exports, "generateModel", { enumerable: true, get: function () { return code_gen_lib_1.generateModel; } });
Object.defineProperty(exports, "generateNamespace", { enumerable: true, get: function () { return code_gen_lib_1.generateNamespace; } });
Object.defineProperty(exports, "ReferenceMaker", { enumerable: true, get: function () { return code_gen_lib_1.ReferenceMaker; } });
var generateAppDeclarationFile_1 = require("./generateAppDeclarationFile");
Object.defineProperty(exports, "generateAppDeclarationFile", { enumerable: true, get: function () { return __importDefault(generateAppDeclarationFile_1).default; } });
var generateRPCProxy_1 = require("./code-gen/generateRPCProxy");
Object.defineProperty(exports, "generateRPCProxy", { enumerable: true, get: function () { return __importDefault(generateRPCProxy_1).default; } });
var generateParser_1 = require("./code-gen/parsers/generateParser");
Object.defineProperty(exports, "generateModelParser", { enumerable: true, get: function () { return __importDefault(generateParser_1).default; } });
Object.defineProperty(exports, "generateNonModelParser", { enumerable: true, get: function () { return generateParser_1.generateNonModelParser; } });
var parsePheroApp_1 = require("./parsePheroApp");
Object.defineProperty(exports, "parsePheroApp", { enumerable: true, get: function () { return parsePheroApp_1.parsePheroApp; } });
__exportStar(require("./errors"), exports);
__exportStar(require("./RPCResult"), exports);
exports.tsx = __importStar(require("./tsx"));
var VirtualCompilerHost_1 = require("./VirtualCompilerHost");
Object.defineProperty(exports, "VirtualCompilerHost", { enumerable: true, get: function () { return VirtualCompilerHost_1.VirtualCompilerHost; } });
