// @phero/core/dist/VirtualCompilerHost.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualCompilerHost = void 0;
const typescript_1 = __importDefault(require("typescript"));
class VirtualCompilerHost {
    constructor(opts) {
        this.files = {};
        this.compilerOpts = {
            ...opts,
            declaration: true,
            // NOTE: we need Promise support in our declaration file. In a normal TS project you would add
            // the "es2015". Because we're implementing a file system here, sort of, we need to set the file
            // name more explicitly. (Implementing our own fileExists makes compilation much much faster.)
            lib: [
                // support for Promise
                "lib.es2015.d.ts",
                // support for Pick, Omit, and other TS utilities
                "lib.es5.d.ts",
            ],
            target: typescript_1.default.ScriptTarget.ES5,
            module: typescript_1.default.ModuleKind.CommonJS,
        };
        const host = typescript_1.default.createCompilerHost(this.compilerOpts);
        host.writeFile = (fileName, contents) => {
            this.files[fileName] = contents;
        };
        host.resolveModuleNames = (moduleNames, containingFile, reusedNames, redirectedReference, options) => {
            return moduleNames.map((moduleName) => {
                // moduleName './kaas' -> fileName "kaas.ts"
                const fileName = `${moduleName.substring("./".length)}.ts`;
                return {
                    resolvedFileName: fileName,
                    isExternalLibraryImport: false,
                };
            });
        };
        const originalReadFile = host.readFile;
        host.readFile = (fileName) => {
            if (this.files[fileName]) {
                return this.files[fileName];
            }
            // Reads the lib files
            return originalReadFile(fileName);
        };
        host.fileExists = (fileName) => {
            return !!this.files[fileName];
        };
        this.host = host;
    }
    addFile(fileName, source) {
        this.files[fileName] = source;
    }
    getFile(fileName) {
        return this.files[fileName];
    }
    createProgram(rootFileNames) {
        return typescript_1.default.createProgram(typeof rootFileNames === "string" ? [rootFileNames] : rootFileNames, this.compilerOpts, this.host);
    }
}
exports.VirtualCompilerHost = VirtualCompilerHost;
